### PR TITLE
Fixed installing bundled dependencies of scoped packages

### LINKED ISSUE #13528: ENOENT errors, `.staging`?
```
npm WARN enoent ENOENT: no such file or directory, open '/root/dev/src/github.com/segment-sources/zendesk/node_modules/esrecurse/package.json'
npm WARN enoent ENOENT: no such file or directory, open '/root/dev/src/github.com/segment-sources/zendesk/node_modules/fs-plus/package.json'
npm WARN enoent ENOENT: no such file or directory, open '/root/dev/src/github.com/segment-sources/zendesk/node_modules/grpc/package.json'
npm WARN enoent ENOENT: no such file or directory, open '/root/dev/src/github.com/segment-sources/zendesk/node_modules/doctrine/package.json'
npm WARN @segment/eslint-config@3.2.1 requires a peer of eslint-plugin-react@^4.0.0 but none was installed.
npm ERR! Linux 4.4.15-moby
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "i"
npm ERR! node v6.3.1
npm ERR! npm  v3.10.3
npm ERR! path /root/dev/src/github.com/segment-sources/zendesk/node_modules/.staging/mkdirp-b0f0282a
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall rename

npm ERR! enoent ENOENT: no such file or directory, rename '/root/dev/src/github.com/segment-sources/zendesk/node_modules/.staging/mkdirp-b0f0282a' -> '/root/dev/src/github.com/segment-sources/zendesk/node_modules/@segment/source/node_modules/grpc/node_modules/node-pre-gyp/node_modules/mkdirp'
npm ERR! enoent ENOENT: no such file or directory, rename '/root/dev/src/github.com/segment-sources/zendesk/node_modules/.staging/mkdirp-b0f0282a' -> '/root/dev/src/github.com/segment-sources/zendesk/node_modules/@segment/source/node_modules/grpc/node_modules/node-pre-gyp/node_modules/mkdirp'
npm ERR! enoent This is most likely not a problem with npm itself
npm ERR! enoent and is related to npm not being able to find a file.
npm ERR! enoent
```

full output- https://gist.github.com/tejasmanohar/ed2497ed7d35d37d9b485134d2a3f1a3
npm debug log- https://gist.github.com/tejasmanohar/93f63476e7bed1bba92c81dc083d2bec

npm -v -> 3.10.3
node -v -> v6.3.1
*nix, ubuntu 16, docker

### LINKED ISSUE #14311: npm@3.5.0 to current fails to install scoped package with bundled dependencies
#### I'm opening this issue because:
- [x] npm is crashing.
#### What's going wrong?

When attempting to install a scoped package with a bundled dependency npm errors. See below for the output on the command line from attempting to install such a package, and here is the debug log: [download](https://github.com/npm/npm/files/534949/npm-debug.log.txt) or [gist](https://gist.github.com/NatalieWolfe/450bef9603cbfb88511d7f1477227bb7).

I have traced the issue down to commit aed71fb9. In particular I believe it is [this line of the diff](https://github.com/npm/npm/commit/aed71fb#diff-6effc31b3479dd85e58db2f9c50dfe2fR58) that is causing the issue. It assumes that the staging directory is in the parent of the build path, however in the case of a scoped module the staging directory should be in the grandparent of the build path. The issue could also stem from [this line instead](https://github.com/npm/npm/commit/aed71fb#diff-6effc31b3479dd85e58db2f9c50dfe2fR30).
#### How can the CLI team reproduce the problem?

Using `npm` version `>=3.5.0` try `npm install @n.wolfe/scoped-test@1.1.0`.

Alternately:
1. Create a scoped package (e.g. `@foo/bar`).
2. Add a dependency to its `package.json` (e.g. `npm install --save express`).
3. Make that dependency bundled.
4. Publish your package (`npm publish --access public`)
5. Attempt to install your package using `npm >=3.5.0` (`npm install @foo/bar`).
### supporting information:

``` sh
$ npm --version
3.10.8
$ node --version
v6.8.1
$ npm config get registry
https://registry.npmjs.org/
$ ls
$ npm install @n.wolfe/scoped-test@1.0.0 # <-- Dependency is not bundled.
/Users/nwolfe/work/me/test
└─┬ @n.wolfe/scoped-test@1.0.0
  └─┬ concat-stream@1.5.2
    ├── inherits@2.0.3
    ├─┬ readable-stream@2.0.6
    │ ├── core-util-is@1.0.2
    │ ├── isarray@1.0.0
    │ ├── process-nextick-args@1.0.7
    │ ├── string_decoder@0.10.31
    │ └── util-deprecate@1.0.2
    └── typedarray@0.0.6

npm WARN enoent ENOENT: no such file or directory, open '/Users/nwolfe/work/me/test/package.json'
npm WARN test No description
npm WARN test No repository field.
npm WARN test No README data
npm WARN test No license field.
$ rm -r node_modules/
$ npm install @n.wolfe/scoped-test@1.1.0 # <-- Dependency IS bundled.
/Users/nwolfe/work/me/test
└── (empty)

npm WARN enoent ENOENT: no such file or directory, open '/Users/nwolfe/work/me/test/package.json'
npm WARN test No description
npm WARN test No repository field.
npm WARN test No README data
npm WARN test No license field.
npm ERR! Darwin 15.6.0
npm ERR! argv "/Users/nwolfe/.nvm/versions/node/v6.8.1/bin/node" "/Users/nwolfe/.nvm/versions/node/v6.8.1/bin/npm" "install" "@n.wolfe/scoped-test@1.1.0"
npm ERR! node v6.8.1
npm ERR! npm  v3.10.8
npm ERR! path /Users/nwolfe/work/me/test/node_modules/.staging/core-util-is-ca852e23
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall rename

npm ERR! enoent ENOENT: no such file or directory, rename '/Users/nwolfe/work/me/test/node_modules/.staging/core-util-is-ca852e23' -> '/Users/nwolfe/work/me/test/node_modules/@n.wolfe/scoped-test/node_modules/core-util-is'
npm ERR! enoent ENOENT: no such file or directory, rename '/Users/nwolfe/work/me/test/node_modules/.staging/core-util-is-ca852e23' -> '/Users/nwolfe/work/me/test/node_modules/@n.wolfe/scoped-test/node_modules/core-util-is'
npm ERR! enoent This is most likely not a problem with npm itself
npm ERR! enoent and is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/nwolfe/work/me/test/npm-debug.log
npm ERR! code 1
```
- Windows, OS X/macOS, or Linux?: Mac OS 10.11.6
- Network issues:
  - Geographic location where npm was run: Portland, OR
  - no proxy or VPN involved
- Container: none

### PR DESCRIPTION
**Pull request by NatalieWolfe**
Added extra steps to resolve staging directory which are required for handling the bundled dependencies of scoped packages. I have confirmed this resolves the issue reported in #14311 and I believe it would fix #13528 as well.

Fixed #14311 
Fixed #13528

### CODE REVIEW

File: lib/install/action/extract.js [hunk 1]
```diff
@@ -28,7 +30,16 @@ function andUpdatePackageJson (pkg, buildpath, next) {
...
 function andStageBundledChildren (pkg, buildpath, log, next) {
-  var staging = path.resolve(buildpath, '..')
+  // Dependencies bundled inside a scoped module need an extra step up when
+  // resolving the path to the staging directory.
+  var staging = buildpath
+  do {
+    staging = path.resolve(staging, '..')
+  } while (staging.length > 1 && path.basename(staging) !== '.staging')
+  if (staging.length <= 1) {
+    throw new Error('???')
```
Comment (NatalieWolfe) [inline lib/install/action/extract.js]: Note: I'm not sure how npm handles errors. Will update this line once I know the correct thing to do. :)

Comment (iarna) [inline lib/install/action/extract.js]: Iit looks like you want a fatal error, for that you'd call the `next` callback with an error object as its first argument.

Output is mediated via `lib/utils/error-message.js`

**Comment by iarna** [comment]
So I've had a rewrite of the argument handling for actions kicking around for ages now that may fix this in a more comprehensive way (by affirmatively passing in the staging directory instead of trying to intuit it from the build directory).

https://github.com/npm/npm/compare/iarna/action-arg-rewrite

I'm going to cherry-pick your test over there and see if it's passing…

**Comment by NatalieWolfe** [comment]
@iarna Sounds promising! Please let me know if that works or if there is anything I can do to help prepare that branch.

**Comment by iarna** [comment]
And yeah, it looks like that does it. Excellent, this gives us a reason to actually merge that. =D

I'm going to close this as I've PRed that branch over in #14403 along with a slightly modified version of your test.

**Pull request closed without merging**